### DISCUSSION TITLE
Why onEach instead of collect {}

### DISCUSSION THREAD

**Discussion by mobilekosmos** [android/nowinandroid#846] [Q&A]
lifecycleScope.launch {
            lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.uiState
                    .onEach {
                        uiState = it
                    }
                    .collect()
            }
        }

I'm trying to learn best practices and I'm not sure why you are using in the MainActivity .onEach instead of just simply using collect {} directly, just personal preference?

**Comment by vamsi3** [comment]
I'm curious about this too.

I did find a [session](https://www.youtube.com/watch?v=fSB6_KE95bU&t=687s) from the Android Dev Summit 2021 where they mention the use of `.collect {...}`.

---

Also, I stumbled upon this discussion post while wondering why we don't use `.collectLatest {...}` itself here, _i.e._,

```kt
lifecycleScope.launch {
    lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED) {
        viewModel.uiState
            .collectLatest { uiState = it }
    }
}
```

Hoping for anyone with more experience in Kotlin Flows to comment on whether this is a valid (and better) code.

**Comment by ParkJong-Hun** [comment]
I think it's written that way when he considers a lot of maintenance of functions.  
Because no mapping method can come after `collect`.

**Comment by anhtuannd** [comment]
@ParkJong-Hun 
Agree with you, onEach is triggered **before** value is emitted downstream. So, it will be applied to all consumers.    
We can also add catch() to handle exception during transmitted.
